cmd/otk-resolve-containers/main_test.go

```diff
@@ -0,0 +1,194 @@
+package main_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/osbuild/images/internal/common"
+	"github.com/osbuild/images/internal/testregistry"
+	"github.com/osbuild/images/pkg/arch"
+	"github.com/osbuild/images/pkg/blueprint"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	resolver "github.com/osbuild/images/cmd/otk-resolve-containers"
+)
+
+const rootLayer = `H4sIAAAJbogA/+SWUYqDMBCG53lP4V5g9x8dzRX2Bvtc0VIhEIhKe/wSKxgU6ktjC/O9hMzAQDL8
+/8yltdb9DLeB0gEGKhHCg/UJsBAL54zKFBAC54ZzyrCUSMfYDydPgHfu6R/s5VePilOfzF/of/bv
+vG2+lqhyFNGPddP53yjyegCBKcuNROZ77AmBoP+CmbIyqpEM5fqf+3/ubJtsCuz7P1b+L1Du/4f5
+v+vrsVPu/Vq9P3ANk//d+x/MZv8TKNf/Qfqf9v9v5fLXK3/lKEc5ypm4AwAA//8DAE6E6nIAEgAA
+`
+
+const (
+	owner    = "osbuild"
+	reponame = "testcontainer"
+)
+
+func createTestRegistry() (*testregistry.Registry, []string) {
+	registry := testregistry.New()
+	repo := registry.AddRepo(fmt.Sprintf("%s/%s", owner, reponame))
+	ref := registry.GetRef(fmt.Sprintf("%s/%s", owner, reponame))
+
+	// add 10 images, all in the same repository with the same content
+	// (rootLayer), but each with a different tag and comment
+	refs := make([]string, 1)
+	for idx := 0; idx < len(refs); idx++ {
+		checksum := repo.AddImage(
+			[]testregistry.Blob{testregistry.NewDataBlobFromBase64(rootLayer)},
+			[]string{"amd64", "ppc64le"},
+			fmt.Sprintf("image %d", idx),
+			time.Time{})
+
+		tag := fmt.Sprintf("tag-%d", idx)
+		repo.AddTag(checksum, tag)
+		refs[idx] = fmt.Sprintf("%s:%s", ref, tag)
+	}
+	return registry, refs
+}
+
+func TestResolver(t *testing.T) {
+	registry, refs := createTestRegistry()
+	defer registry.Close()
+
+	inpContainers := make([]blueprint.Container, len(refs))
+	for idx, ref := range refs {
+		inpContainers[idx] = blueprint.Container{
+			Source:       ref,
+			Name:         fmt.Sprintf("test/localhost/%s", ref), // add a prefix for the local name to override the source
+			TLSVerify:    common.ToPtr(false),
+			LocalStorage: false,
+		}
+	}
+
+	for _, containerArch := range []string{"amd64", "ppc64le"} {
+		t.Run(containerArch, func(t *testing.T) {
+
+			require := require.New(t)
+			assert := assert.New(t)
+
+			amd64input := map[string]interface{}{
```

Comment:
I may sound silly here, but I'm a fan of writing the json out mostly because to me a test is also documentation and having a look into the test file to see what the json input and expected json output looks like feels rather nice. I can see it's tedious and work so feel free to ignore (both there and in the expected output)

Reply:
In this case it's incredibly tedious but for the output it's also mostly impossible to get right because the order or elements (and sometimes keys) isn't predictable.  It's often stable, and the elements are stable in our container resolver, but modifying the tests in any way would require adjusting the expected output and figuring out the new order of elements.

Not to mention, the error output if it fails is a lot less readable when it's just comparing json in string format.

Reply:
I just noticed I accidentally left the test running with only 1 ref when the comment says 10 (it should be 10).  But I'll add this test with raw json and a single ref too.